pkg/scale/encode.go

```diff
@@ -0,0 +1,357 @@
+package scale
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"reflect"
+)
+
+func Marshal(v interface{}) (b []byte, err error) {
+	es := encodeState{
+		fieldScaleIndicesCache: cache,
+	}
+	err = es.marshal(v)
+	if err != nil {
+		return
+	}
+	b = es.Bytes()
+	return
+}
+
+type encodeState struct {
+	bytes.Buffer
+	*fieldScaleIndicesCache
+}
+
+func (es *encodeState) marshal(in interface{}) (err error) {
+	switch in := in.(type) {
+	case int, uint, int8, uint8, int16, uint16, int32, uint32, int64, uint64:
+		err = es.encodeFixedWidthInt(in)
+	case *big.Int:
+		err = es.encodeBigInt(in)
+	case Uint128:
+		err = es.encodeUint128(in)
+	case []byte:
+		err = es.encodeBytes(in)
+	case string:
+		err = es.encodeBytes([]byte(in))
+	case bool:
+		err = es.encodeBool(in)
+	case VaryingDataType:
+		err = es.encodeVaryingDataType(in)
+	// TODO: case common.Hash:
+	// 	n, err = es.Writer.Write(v.ToBytes())
+	default:
+		switch reflect.TypeOf(in).Kind() {
+		case reflect.Ptr:
+			// Assuming that anything that is a pointer is an Option to capture {nil, T}
+			elem := reflect.ValueOf(in).Elem()
+			switch elem.IsValid() {
+			case false:
+				err = es.WriteByte(0)
+			default:
+				err = es.WriteByte(1)
+				if err != nil {
+					return
+				}
+				err = es.marshal(elem.Interface())
+			}
+		case reflect.Struct:
+			err = es.encodeStruct(in)
+		case reflect.Array:
+			err = es.encodeArray(in)
+		case reflect.Slice:
+			err = es.encodeSlice(in)
+		default:
+			err = fmt.Errorf("unsupported type: %T", in)
+		}
+	}
+	return
+}
+
+func (es *encodeState) encodeVaryingDataType(values VaryingDataType) (err error) {
+	// err = es.encodeLength(len(types))
+	err = es.encodeLength(len(values))
+	if err != nil {
+		return
+	}
+	for _, t := range values {
+		// encode type.Index (idx) for varying data type
+		err = es.WriteByte(byte(t.Index()))
+		if err != nil {
+			return
+		}
+		err = es.marshal(t)
+	}
+	return
+}
+
+func (es *encodeState) encodeSlice(t interface{}) (err error) {
+	switch arr := t.(type) {
+	// int is the only case that handles encoding differently.
+	// all other cases can recursively call es.marshal()
+	case []int:
+		err = es.encodeLength(len(arr))
+		if err != nil {
+			return
+		}
+		for _, elem := range arr {
+			err = es.encodeUint(uint(elem))
```

Comment:
if there some signed integer in `arr` then we can have an overflow, maybe we could create a new case type to `[]uint`.

Reply:
it's not clearly defined in the spec or the knowledge base but it's assumed that all these are unsigned integers.

Reply:
yes this is fine, when encoding all ints are treated as uints, the user would just need to decode it into the right type